README.md
# Vector Align

This is a production-ready tool for align parallel documents to sentences without the need for a machine translation system or lexicon.
[Vecalign](https://github.com/thompsonb/vecalign) - an accurate sentence alignment algorithm is used, which is fast even for very long documents.

Instead of [LASER](https://github.com/facebookresearch/LASER), we use [LaBSE](https://arxiv.org/abs/2007.01852) from [sentence-transformers](https://github.com/UKPLab/sentence-transformers) pre-trained models for its better performance in parallel text mining task.

To support the specific African languages in WMT22, we fork from [laserembeddings](https://github.com/yannvgn/laserembeddings) and integrate sentence pieces tokenizer in [this repository](https://github.com/volctrans/laserembeddings).

# Installation (Recommend)
```Bash
pip3 setup.py install 
```


# Example
1. two separate docs
```bash
vector-align \
    --left-doc tests/zh.file \
    --right-doc tests/en.file \
    --output output.txt
```

2. 1 bilingual doc
```bash
vector-align \
    --left-doc tests/bilingual.file \
    --right-doc tests/bilingual.file \
    --output output.txt
```

# Reference
Please cite the [paper](https://www.aclweb.org/anthology/D19-1136.pdf) if you use this tool:

```
@inproceedings{thompson-koehn-2019-vecalign,
    title = "{V}ecalign: Improved Sentence Alignment in Linear Time and Space",
    author = "Thompson, Brian and Koehn, Philipp",
    booktitle = "Proceedings of the 2019 Conference on Empirical Methods in Natural Language Processing and the 9th International Joint Conference on Natural Language Processing (EMNLP-IJCNLP)",
    month = nov,
    year = "2019",
    address = "Hong Kong, China",
    publisher = "Association for Computational Linguistics",
    url = "https://www.aclweb.org/anthology/D19-1136",
    doi = "10.18653/v1/D19-1136",
    pages = "1342--1348",
}
```


CHANGELOG.md
<a name="0.0.1"></a>
# 0.1.1 (2022-04-08)

- Initial version 🐣


tests/__init__.py
# -*- coding: utf-8 -*-
# create@ 2022-04-08 11:18

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals


vector_align/utils.py
# -*- coding: utf-8 -*-
# create@ 2022-04-08 19:44

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import re
from typing import Tuple

import pycld2 as cld2
from polyglot.text import Text

line_break_regex = re.compile(r'(\\n)+\*?')


def group_sentences(content: str, language: str = None, do_split: bool = True) -> Tuple:
    """
    split content into sentences filtered with language, if provided
    :param content:
    :param language:
    :param do_split:
    :return:
    """
    content = normalize(content)

    if language is None:
        reliable, _, top_3_choices = cld2.detect(content, bestEffort=False)
        language = top_3_choices[0][1]

    if do_split:

        collection = []
        lines = content.split('\n')
        for line in lines:
            t = Text(line, language)
            for sent in t.sentences:
                raw = sent.raw
                reliable, _, top_3_choices = cld2.detect(raw, bestEffort=False)
                if top_3_choices:
                    top_lang, top_lang_percent = top_3_choices[0][1], top_3_choices[0][2]
                    if top_lang_percent > 80:
                        if top_lang == language:
                            collection.append(raw)
                else:
                    collection.append(raw)
    else:
        collection = content.split('\n')
    return collection, language


def group_sentences_in_pair(content: str, src_lang: str = None, trg_lang: str = None,
                            do_split: bool = False) -> Tuple:
    """
    split content into sentences filtered with language, if provided
    :param content:
    :param src_lang:
    :param trg_lang:
    :param do_split:
    :return:
    """
    content = normalize(content)
    reliable, _, top_3_choices = cld2.detect(content, bestEffort=False)
    src_collection = []
    trg_collection = []
    if len(top_3_choices) >= 2:
        top_1_lang, top_2_lang = min(top_3_choices[0][1], top_3_choices[1][1]), \
                                 max(top_3_choices[0][1], top_3_choices[1][1])

        if src_lang and trg_lang:
            assert (top_1_lang, top_2_lang) == (min(src_lang, trg_lang), max(src_lang, trg_lang))

        else:
            src_lang, trg_lang = top_1_lang, top_2_lang
        lines = content.split('\n')
        for line in lines:
            if do_split:
                t = Text(line)
                sentences = [sent.raw for sent in t.sentences]
            else:
                sentences = [line]
            for raw in sentences:
                reliable, _, top_3_choices = cld2.detect(raw, bestEffort=False)
                if top_3_choices:
                    top_lang, top_lang_percent = top_3_choices[0][1], top_3_choices[0][2]
                    if top_lang_percent > 80:
                        if top_lang == src_lang:
                            src_collection.append(raw)
                        else:
                            trg_collection.append(raw)
    return src_collection, trg_collection, src_lang, trg_lang


def normalize(text):
    # remove multiple line break character
    doc = line_break_regex.sub('\n', text)
    lines = []
    for line in doc.split('\n'):
        line = line.strip().lstrip('*').lstrip('-')
        if len(line) > 1:
            lines.append(line)
    doc = "\n".join(lines)
    return doc


vector_align/extractor.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import os
import re
import zipfile
from math import ceil
from pathlib import Path
from typing import List

import numpy as np
import requests
from sentence_transformers import SentenceTransformer

from . import logger
from .dp_utils import yield_overlaps, make_alignment_types, make_doc_embedding, vecalign

CUR_DIR = os.path.abspath(os.path.dirname(__file__))

home = os.path.expanduser('~')

DATA_DIR = os.path.join(home, f'.cache/transformers/')
if not os.path.exists(DATA_DIR):
    os.makedirs(DATA_DIR, exist_ok=True)
MODEL_DIR = os.path.join(DATA_DIR, 'LaBSE')
MODEL_FILE = os.path.join(DATA_DIR, 'LaBSE.zip')

if not Path(MODEL_DIR).exists():
    model_url = 'http://tosv.byted.org/obj/lab-mt-data/sentence-transformers/LaBSE.zip'
    try:
        logger.info('= Begin download sentence-transformers model. =')
        response = requests.get(model_url, stream=True)
        total_length = int(response.headers.get('content-length'))
        with open(MODEL_FILE, "wb") as f:
            dl = 0
            for chunk in response.iter_content(chunk_size=20480):
                if chunk:
                    dl += len(chunk)
                    f.write(chunk)
                    done = int(50 * dl / total_length)
                    logger.info(f"Download:[{'=' * done}{' ' * (50 - done)}]")
        with zipfile.ZipFile(MODEL_FILE, 'r') as zip_ref:
            zip_ref.extractall(DATA_DIR)
        Path(MODEL_FILE).unlink()
    except Exception as ex:
        logger.exception(ex)

LaBSE_model = SentenceTransformer(model_name_or_path=MODEL_DIR)


def overlap(sentences_group: List[str], num_overlaps: int) -> List[str]:
    output = set()
    for out_line in yield_overlaps(sentences_group, num_overlaps):
        output.add(out_line)

    # for reproducibility
    output = list(output)
    output.sort()
    return output


def embedding(sentences_group: List[str]):
    return LaBSE_model.encode(sentences_group)


def clean(sentences_group: List[str]) -> List[str]:
    return [re.sub(re.compile(r"[\s]+"), " ", text).strip() for text in sentences_group]


def read_in_embeddings(sentences_group_overlap: List[str], embed: np.ndarray):
    """
    Given a text file with candidate sentences and a corresponing embedding file,
       make a maping from candidate sentence to embedding index, 
       and a numpy array of the embeddings
    """
    sent2line = dict()
    for ii, line in enumerate(sentences_group_overlap):
        if line.strip() in sent2line:
            raise Exception('got multiple embeddings for the same line')
        sent2line[line.strip()] = ii

    return sent2line, embed


def extract_parallel_pairs(src_sentences_group: List[str],
                           trg_sentences_group: List[str],
                           alignment_max_size: int = 4,
                           del_percentile_frac: float = 0.2,
                           max_size_full_dp: int = 300,
                           costs_sample_size: int = 20000,
                           num_samps_for_norm: int = 100,
                           search_buffer_size: int = 5) -> List:
    alignment_max_size = max(alignment_max_size, 2)

    src_sentences_group = clean(src_sentences_group)
    trg_sentences_group = clean(trg_sentences_group)

    src_sentences_group_overlap = overlap(src_sentences_group, num_overlaps=alignment_max_size)
    trg_sentences_group_overlap = overlap(trg_sentences_group, num_overlaps=alignment_max_size)

    src_sentences_group_emb = embedding(src_sentences_group_overlap)
    trg_sentences_group_emb = embedding(trg_sentences_group_overlap)

    src_sent2line, src_line_embeddings = read_in_embeddings(src_sentences_group_overlap,
                                                            src_sentences_group_emb)
    tgt_sent2line, tgt_line_embeddings = read_in_embeddings(trg_sentences_group_overlap,
                                                            trg_sentences_group_emb)

    width_over2 = ceil(alignment_max_size / 2.0) + search_buffer_size

    src_lines = src_sentences_group
    vecs0 = make_doc_embedding(src_sent2line, src_line_embeddings, src_lines, alignment_max_size)

    tgt_lines = trg_sentences_group
    vecs1 = make_doc_embedding(tgt_sent2line, tgt_line_embeddings, tgt_lines, alignment_max_size)

    final_alignment_types = make_alignment_types(alignment_max_size)

    stack = vecalign(vecs0=vecs0,
                     vecs1=vecs1,
                     final_alignment_types=final_alignment_types,
                     del_percentile_frac=del_percentile_frac,
                     width_over2=width_over2,
                     max_size_full_dp=max_size_full_dp,
                     costs_sample_size=costs_sample_size,
                     num_samps_for_norm=num_samps_for_norm)

    res = []
    for (x, y), s in zip(stack[0]['final_alignments'], stack[0]['alignment_scores']):
        if len(x) == 0 or len(y) == 0 or s >= 0.5:
            continue
        src = ''
        for i in x:
            src += src_sentences_group[i]
        tgt = ''
        for i in y:
            tgt += trg_sentences_group[i]
        res.append((src, tgt, s))
    return res


vector_align/main.py
# -*- coding: utf-8 -*-
# create@ 2022-04-08 19:18

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import argparse
import codecs
import json
from pathlib import Path

from vector_align.extractor import extract_parallel_pairs
from vector_align.utils import group_sentences, group_sentences_in_pair


def main():
    parser = argparse.ArgumentParser("Vector Align Tool")

    parser.add_argument("--left-doc",
                        type=str,
                        required=True)

    parser.add_argument("--right-doc",
                        type=str,
                        required=True)

    parser.add_argument("--left-lang",
                        type=str,
                        required=False)

    parser.add_argument("--right-lang",
                        type=str,
                        required=False)

    parser.add_argument("--method",
                        type=str,
                        required=False,
                        choices=['labse', 'laser'],
                        default="labse")

    parser.add_argument("--output",
                        type=str,
                        required=False)

    args = parser.parse_args()

    left_doc = Path(args.left_doc)
    right_doc = Path(args.right_doc)

    assert left_doc.is_file() and right_doc.is_file(), "Provided left doc/right doc is invalid."

    if args.left_doc == args.right_doc:
        with codecs.open(args.left_doc, errors='ignore') as f:
            content = f.read()
            left_sentences, right_sentences, left_lang, right_lang = \
                group_sentences_in_pair(content,
                                        args.left_lang,
                                        args.right_lang)
    else:
        with codecs.open(args.left_doc, errors='ignore') as lf, \
                codecs.open(args.right_doc, errors='ignore') as rf:
            left_content, right_content = lf.read(), rf.read()
            left_sentences, left_lang = group_sentences(left_content, args.left_lang)
            right_sentences, right_lang = group_sentences(right_content, args.right_lang)
    extracted_pairs = extract_parallel_pairs(left_sentences, right_sentences)

    with open(args.output or 'output.txt', 'w') as f:
        for sentence_pair in extracted_pairs:
            line = {
                'src_text': sentence_pair[0],
                'trg_text': sentence_pair[1],
                'src_lang': left_lang,
                'trg_lang': right_lang,
                'score': round(1 - sentence_pair[2], 2)
            }
            f.write(json.dumps(line, ensure_ascii=False) + '\n')


if __name__ == '__main__':
    main()
